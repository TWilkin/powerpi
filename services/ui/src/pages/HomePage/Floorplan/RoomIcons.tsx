import { Room } from "@powerpi/common-api";
import { useCallback, useMemo } from "react";
import { useTranslation } from "react-i18next";
import { chain as _ } from "underscore";
import DeviceIcon from "../../../components/DeviceIcon";
import Icon from "../../../components/Icon";
import SensorIcon from "../../../components/SensorIcon";
import { isPolygonRoom } from "./ViewBox";
import useRoomDevices from "./useRoomDevices";

type RoomIconsProps = {
    room: Room;
};

/** Component to display device/sensor icons in a room on the floorplan. */
const RoomIcons = ({ room }: RoomIconsProps) => {
    const { t } = useTranslation();

    const devices = useRoomDevices(room.name);

    const { scale, iconPaddedSize, deviceCount, iconsWide, offsetX, offsetY } = useMemo(() => {
        const iconSize = 40;
        const iconPadding = 4;
        const iconPaddedSize = iconSize + 2 * iconPadding;

        const scale = 1 / (1024 / iconSize);

        // calculate the room size
        let x = room.x ?? 0;
        let y = room.y ?? 0;
        let width = 0;
        let height = 0;

        // this could put the icons outside the polygon so needs to be improved
        if (isPolygonRoom(room)) {
            x = Math.min(...room.points.map((point) => point.x));
            y = Math.min(...room.points.map((point) => point.y));

            width = Math.max(...room.points.map((point) => point.x)) - x;
            height = Math.max(...room.points.map((point) => point.y)) - y;
        } else {
            width = room.width ?? 0;
            height = room.height ?? 0;
        }

        // next we need to find the centre of the room
        const centreX = width / 2 + x;
        const centreY = height / 2 + y;

        // how many icons can we fit?
        let iconsWide = Math.floor(width / iconPaddedSize);
        let iconsTall = Math.floor(height / iconPaddedSize);

        // we want to leave some space either side
        iconsWide = Math.max(iconsWide - 2, 0);
        iconsTall = Math.max(iconsTall - 2, 0);

        // limit them to what is necessary to centre everything
        const best = bestValue(devices.length, iconsWide, iconsTall);
        if (best) {
            iconsWide = best[0];
            iconsTall = best[1];
        }

        // how many devices can we actually show
        let deviceCount = Math.min(devices.length, iconsWide * iconsTall);
        if (deviceCount < devices.length) {
            // if we have too many devices we'll show an ellipsis instead of the last one
            deviceCount--;
        }

        // now we can work out the offsets
        const offsetX = centreX - (iconsWide * iconPaddedSize) / 2;
        const offsetY = centreY - (iconsTall * iconPaddedSize) / 2;

        return {
            scale,
            iconPaddedSize,
            deviceCount,
            iconsWide,
            offsetX,
            offsetY,
        };
    }, [devices.length, room]);

    const getTransform = useCallback(
        (index: number) => {
            // find out which row and column it's in
            let row = 0;
            while (index >= (row + 1) * iconsWide) {
                row++;
            }
            const column = index - row * iconsWide;

            // now we can work out the position
            const x = offsetX + column * iconPaddedSize;
            const y = offsetY + row * iconPaddedSize;

            return `translate(${x} ${y}) scale(${scale})`;
        },
        [iconPaddedSize, iconsWide, offsetX, offsetY, scale],
    );

    return (
        <>
            {_(devices)
                .first(deviceCount)
                .map((device, i) => (
                    <g key={`${device.deviceType}_${device.type}`} transform={getTransform(i)}>
                        <title>
                            {device.type} ({device.count})
                        </title>

                        {device.deviceType === "device" && <DeviceIcon type={device.type} />}

                        {device.deviceType === "sensor" && (
                            <SensorIcon type={device.type} state={device.state} />
                        )}
                    </g>
                ))
                .value()}

            {deviceCount < devices.length && (
                <g transform={getTransform(deviceCount)}>
                    <title>
                        {t("pages.home.more devices", { count: devices.length - deviceCount })}
                    </title>

                    <Icon icon="more" />
                </g>
            )}
        </>
    );
};
export default RoomIcons;

/** Generate factors for value starting with the values closes to the square root. */
function* factors(value: number) {
    const range = Math.floor(Math.sqrt(value));

    for (let i = range; i >= 1; i--) {
        if (value % i === 0) {
            yield [i, value / i];
        }
    }
}

/** Generate factors between count and rows * columns, in case the factors of count won't fit. */
function* increasingFactors(count: number, rows: number, columns: number) {
    const maxCount = Math.min(count, rows * columns);

    for (let i = maxCount; i <= rows * columns; i++) {
        for (const f of factors(i)) {
            yield f;
        }
    }
}

/** Generate factors of count that will work for the rows and columns. */
function* feasibleFactors(count: number, rows: number, columns: number) {
    for (const [f1, f2] of increasingFactors(count, rows, columns)) {
        if (f1 <= columns && f2 <= rows) {
            yield [f2, f1];
        } else if (f2 <= columns && f1 <= rows) {
            yield [f1, f2];
        }
    }
}

/** Find the "best" factors of count in a grid of no more than rows x columns. */
function bestValue(count: number, rows: number, columns: number) {
    for (const factors of feasibleFactors(count, rows, columns)) {
        return factors;
    }

    return undefined;
}
