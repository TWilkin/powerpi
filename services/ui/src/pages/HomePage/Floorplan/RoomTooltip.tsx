import { Sensor } from "@powerpi/common-api";
import classNames from "classnames";
import { useMemo } from "react";
import { Fragment } from "react/jsx-runtime";
import BatteryIcon from "../../../components/BatteryIcon";
import SensorIcon from "../../../components/SensorIcon";
import Tooltip from "../../../components/Tooltip";
import Value from "../../../components/Value";

type RoomTooltipProps = {
    name: string;

    floor: string;

    room: string;

    sensors: Sensor[];
};

const RoomTooltip = ({ name, floor, room, sensors }: RoomTooltipProps) => {
    const showingBattery = useMemo(
        () => sensors.findIndex((sensor) => (sensor.batterySince ?? 0) > 0) !== -1,
        [sensors],
    );

    return (
        <Tooltip id={generateRoomTooltipId(floor, room)} place="top">
            <h3>{name}</h3>

            <div
                className={classNames("grid auto-rows-auto", {
                    "grid-cols-[20px_20px_auto_1fr]": showingBattery,
                    "grid-cols-[20px_auto_1fr]": !showingBattery,
                })}
            >
                {sensors.map((sensor) => (
                    <Fragment key={sensor.name}>
                        <SensorIcon type={sensor.type} state={sensor.state} />

                        {showingBattery &&
                            (sensor.battery != null ? <BatteryIcon device={sensor} /> : <span />)}

                        <p className="px-2">{sensor.display_name}</p>

                        <p>
                            <Value type={sensor.type} value={sensor.value} unit={sensor.unit} />
                        </p>
                    </Fragment>
                ))}
            </div>
        </Tooltip>
    );
};
export default RoomTooltip;

export function generateRoomTooltipId(floor: string, room: string) {
    return `room-tooltip-${floor}-${room}`;
}
