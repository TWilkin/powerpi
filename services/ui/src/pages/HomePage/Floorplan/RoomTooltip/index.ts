export { default as generateRoomTooltipId } from "./generateRoomTooltipId";
export { default } from "./RoomTooltip";
