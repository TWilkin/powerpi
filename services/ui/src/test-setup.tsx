import { PowerPiApi } from "@powerpi/common-api";
import {
    RenderHookOptions,
    RenderOptions,
    render as testRender,
    renderHook as testRenderHook,
} from "@testing-library/react";
import { PropsWithChildren, ReactNode } from "react";
import { QueryClient, QueryClientProvider } from "react-query";
import { instance, mock } from "ts-mockito";
import { PowerPiAPIContextProvider } from "./hooks/api";

type CommonProps = {
    api?: PowerPiApi;
    queryClient?: QueryClient;
};

/** Extract our options from those used by `render`/`renderHook` and initialise those that aren't set. */
function setupOptions<TOptionType extends CommonProps>(options: TOptionType) {
    let { api, queryClient, ...defaultOptions } = options;

    api ??= instance(mock<PowerPiApi>());
    queryClient ??= new QueryClient();

    defaultOptions = {
        ...defaultOptions,
        wrapper: ({ children }: PropsWithChildren<unknown>) => (
            <Wrapper api={api} queryClient={queryClient}>
                {children}
            </Wrapper>
        ),
    };

    return defaultOptions;
}

/** Wrap `render` method to setup the API and `react-query`. */
function render(component: ReactNode, options: RenderOptions & CommonProps) {
    const defaultOptions = setupOptions(options);

    return testRender(component, defaultOptions);
}

/** Wrap `renderHook` method to setup the API and `react-query`. */
function renderHook<TResultType>(
    render: Parameters<typeof testRenderHook>[0],
    options: RenderHookOptions<TResultType> & CommonProps,
) {
    const defaultOptions = setupOptions(options);

    return testRenderHook(render, defaultOptions);
}

type WrapperProps = PropsWithChildren<Required<CommonProps>>;

/** The `Wrapper` component simply wraps the component/hook under test with the API and `react-query` providers. */
const Wrapper = ({ api, queryClient, children }: WrapperProps) => (
    <QueryClientProvider client={queryClient}>
        <PowerPiAPIContextProvider api={api}>{children}</PowerPiAPIContextProvider>
    </QueryClientProvider>
);

export * from "@testing-library/react";
export { render, renderHook };
