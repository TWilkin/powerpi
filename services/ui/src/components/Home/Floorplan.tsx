import {
    Floor as IFloor,
    Floorplan as IFloorplan,
    Point as IPoint,
    Room as IRoom,
} from "@powerpi/common-api";
import classNames from "classnames";
import { useMemo } from "react";
import { useParams } from "react-router-dom";
import useOrientation from "../../hooks/orientation";
import styles from "./Floorplan.module.scss";
import RoomIcons from "./RoomIcons";

interface FloorplanProps {
    floorplan: IFloorplan;
}

const Floorplan = ({ floorplan }: FloorplanProps) => {
    const params = useParams();
    const { isLandscape, isPortrait } = useOrientation();

    const size = useMemo(() => viewBoxByFloorplan(floorplan), [floorplan]);

    const isWide = useMemo(() => size.maxX >= size.maxY, [size]);

    const rotate = useMemo(
        () => (isWide && isPortrait) || (!isWide && isLandscape),
        [isWide, isLandscape, isPortrait],
    );

    return (
        <div className={styles.layout}>
            <svg
                viewBox={`${size.minX} ${size.minY} ${size.maxX} ${size.maxY}`}
                preserveAspectRatio="xMidYMid"
                className={classNames({ [styles.rotate]: rotate, [styles.wide]: isWide })}
            >
                <defs>
                    {floorplan.floors.map((floor) => (
                        <RoomOutline key={floor.name} floor={floor} />
                    ))}
                </defs>

                {floorplan.floors.map((floor) => (
                    <Floor
                        key={floor.name}
                        floor={floor}
                        visible={params.floor === floor.name}
                        rotate={rotate}
                    />
                ))}
            </svg>
        </div>
    );
};
export default Floorplan;

const RoomOutline = ({ floor }: { floor: IFloor }) => {
    const size = useMemo(() => viewBoxByFloor(floor), [floor]);

    return (
        <filter
            id={outlineId(floor)}
            x={size.minX}
            y={size.minY}
            width={size.maxX}
            height={size.maxY}
            filterUnits="userSpaceOnUse"
        >
            <feMorphology operator="dilate" in="SourceAlpha" radius={2} result="morph1" />
            <feMorphology operator="dilate" in="SourceAlpha" radius={3} result="morph2" />
            <feComposite in="morph1" in2="morph2" operator="xor" result="outline" />
            <feComposite in="outline" in2="SourceGraphic" operator="over" result="output" />
        </filter>
    );
};

interface FloorProps {
    floor: IFloor;
    visible: boolean;
    rotate: boolean;
}

const Floor = ({ floor, visible, rotate }: FloorProps) => {
    return (
        <g
            id={floor.name}
            filter={`url(#${outlineId(floor)})`}
            className={classNames(styles.floor, { [styles.visible]: visible })}
        >
            <title>{floor.display_name ?? floor.name}</title>

            {floor.rooms.map((room) => (
                <Room key={room.name} room={room} floor={floor.name} rotate={rotate} />
            ))}
        </g>
    );
};

interface RoomProps {
    room: IRoom;
    floor: string;
    rotate: boolean;
}

const Room = ({ room, floor, rotate }: RoomProps) => {
    const id = useMemo(() => `${floor}${room.name}`, [floor, room]);

    if (isRect(room)) {
        return (
            <g id={id} data-tooltip-id={id}>
                <rect x={room.x} y={room.y} width={room.width} height={room.height} />
                <RoomIcons room={room} rotate={rotate} />
            </g>
        );
    } else if (isPolygon(room)) {
        const points = room.points?.map((point) => `${point.x},${point.y}`).join(" ");

        return (
            <g id={id} data-tooltip-id={id}>
                <polygon points={points} />
                <RoomIcons room={room} rotate={rotate} />
            </g>
        );
    }

    return <></>;
};

const outlineId = (floor: IFloor) => `${floor.name}Outline`;

const isRect = (room: IRoom) =>
    room.width && room.height && (!room.points || room.points.length === 0);

const isPolygon = (room: IRoom) =>
    !room.width && !room.height && room.points && room.points.length !== 0;

class ViewBox {
    minX: number;
    minY: number;
    maxX: number;
    maxY: number;

    constructor() {
        this.minX = Number.MAX_VALUE;
        this.minY = Number.MAX_VALUE;
        this.maxX = 0;
        this.maxY = 0;
    }

    pad(size = 5) {
        this.minX -= size;
        this.minY -= size;
        this.maxX += size;
        this.maxY += size;

        return this;
    }

    join(viewBox: ViewBox) {
        this.minX = Math.min(this.minX, viewBox.minX);
        this.minY = Math.min(this.minY, viewBox.minY);
        this.maxX = Math.max(this.maxX, viewBox.maxX);
        this.maxY = Math.max(this.maxY, viewBox.maxY);

        return this;
    }

    update(x: number, y: number) {
        this.minX = Math.min(this.minX, x);
        this.minY = Math.min(this.minY, y);
        this.maxX = Math.max(this.maxX, x);
        this.maxY = Math.max(this.maxY, y);

        return this;
    }
}

const viewBoxByFloorplan = (floorplan: IFloorplan) =>
    floorplan.floors
        .map((floor) => viewBoxByFloor(floor))
        .reduce((viewBox, floor) => viewBox.join(floor), new ViewBox())
        .pad();

const viewBoxByFloor = (floor: IFloor) =>
    floor.rooms
        .map((room) => viewBoxByRoom(room))
        .reduce((viewBox, room) => viewBox.join(room), new ViewBox());

function viewBoxByRoom(room: IRoom) {
    // convert the room into a list of points
    const points: IPoint[] = [];
    if (isRect(room)) {
        points.push(...pointsFromRect(room.x, room.y, room.width, room.height));
    } else if (isPolygon(room)) {
        points.push(...(room.points ?? []));
    }

    // now get the min/max values
    const viewBox = points.reduce(
        (viewBox, point) => viewBox.update(point.x, point.y),
        new ViewBox(),
    );

    return viewBox;
}

function pointsFromRect(x = 0, y = 0, width = 0, height = 0): IPoint[] {
    return [
        { x, y },
        { x: x + width, y: y },
        { x, y: y + height },
        { x: x + width, y: y + height },
    ];
}
