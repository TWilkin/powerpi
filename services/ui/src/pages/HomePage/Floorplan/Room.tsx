import { Room as IRoom } from "@powerpi/common-api";
import classNames from "classnames";
import { useMemo } from "react";
import RoomIcons from "./RoomIcons";
import { generateRoomTooltipId } from "./RoomTooltip";
import { isPolygonRoom, isRectangleRoom } from "./ViewBox";

type RoomProps = {
    floor: string;

    room: IRoom;
};

const shapeClasses = classNames(
    "pointer-events-all",
    "fill-none outline-none",
    "stroke-text",
    "hover:fill-bg-hover",
);

/** Component representing a room in the home floorplan. */
const Room = ({ floor, room }: RoomProps) => {
    const points = useMemo(() => {
        if (isPolygonRoom(room)) {
            return room.points.map((point) => `${point.x},${point.y}`).join(" ");
        }

        return undefined;
    }, [room]);

    return (
        <g className="focus:outline-none" data-tooltip-id={generateRoomTooltipId(floor, room.name)}>
            {isRectangleRoom(room) && (
                <rect
                    x={room.x}
                    y={room.y}
                    width={room.width}
                    height={room.height}
                    className={shapeClasses}
                />
            )}

            {points && <polygon points={points} className={shapeClasses} />}

            <RoomIcons room={room} />
        </g>
    );
};
export default Room;
