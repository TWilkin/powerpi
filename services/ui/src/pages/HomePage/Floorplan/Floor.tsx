import { Floor as IFloor } from "@powerpi/common-api";
import Room from "./Room";

type FloorProps = {
    floor: IFloor;
};

/** Component representing a floor in the home floorplan. */
const Floor = ({ floor }: FloorProps) => (
    <g>
        {floor.rooms.map((room) => (
            <Room key={room.name} room={room} />
        ))}
    </g>
);
export default Floor;
