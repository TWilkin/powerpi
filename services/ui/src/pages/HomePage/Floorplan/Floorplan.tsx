import { Floorplan as IFloorplan } from "@powerpi/common-api";
import { useMemo } from "react";
import useFloor from "../useFloor";
import Floor from "./Floor";

type FloorplanProps = {
    floorplan: IFloorplan;
};

/** Component representing the SVG of the home floorplan. */
const Floorplan = ({ floorplan }: FloorplanProps) => {
    const currentFloorName = useFloor();

    const currentFloor = useMemo(
        () => floorplan.floors.find((floor) => floor.name === currentFloorName),
        [currentFloorName, floorplan.floors],
    );

    return (
        <>
            {currentFloor && (
                <svg preserveAspectRatio="xMidYMid" className="flex-1">
                    <Floor floor={currentFloor} />
                </svg>
            )}
        </>
    );
};
export default Floorplan;
