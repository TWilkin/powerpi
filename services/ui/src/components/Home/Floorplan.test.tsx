import { act, render } from "../../test-setup";
import Floorplan from "./Floorplan";

describe("Floorplan", () => {
    beforeEach(() => {
        Object.defineProperty(window, "matchMedia", {
            writable: true,
            value: jest.fn().mockImplementation((query) => ({
                matches: false,
                media: query,
                onchange: null,

                addEventListener: jest.fn(),
                removeEventListener: jest.fn(),
                dispatchEvent: jest.fn(),
            })),
        });
    });

    test("renders", async () => {
        const { container } = await act(() =>
            render(
                <Floorplan
                    floorplan={{
                        floors: [
                            {
                                name: "Basement",
                                rooms: [
                                    {
                                        name: "Basement",
                                        width: 100,
                                        height: 100,
                                    },
                                ],
                            },
                            {
                                name: "Ground",
                                display_name: "Neverland",
                                rooms: [
                                    {
                                        name: "LivingRoom",
                                        display_name: "Living Room",
                                        points: [
                                            { x: 0, y: 0 },
                                            { x: 50, y: 0 },
                                            { x: 50, y: 100 },
                                            { x: 0, y: 100 },
                                        ],
                                    },
                                    {
                                        name: "Kitchen",
                                        points: [
                                            { x: 50, y: 0 },
                                            { x: 100, y: 0 },
                                            { x: 100, y: 100 },
                                            { x: 50, y: 100 },
                                        ],
                                    },
                                ],
                            },
                        ],
                    }}
                />,
            ),
        );

        expect(container).toMatchSnapshot();
    });
});
