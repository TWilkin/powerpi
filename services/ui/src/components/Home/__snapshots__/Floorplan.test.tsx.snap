// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Floorplan renders 1`] = `
<div>
  <div
    class="layout"
  >
    <svg
      preserveAspectRatio="xMidYMid"
      viewBox="0 0 100 100"
    >
      <g
        class="floor"
        id="Basement"
      >
        <g
          data-tooltip-id="BasementBasement"
          id="BasementBasement"
        >
          <rect
            height="100"
            width="100"
          />
        </g>
      </g>
      <g
        class="floor"
        id="Ground"
      >
        <g
          data-tooltip-id="GroundLivingRoom"
          id="GroundLivingRoom"
        >
          <polygon
            points="0,0 50,0 50,100 0,100"
          />
        </g>
        <g
          data-tooltip-id="GroundKitchen"
          id="GroundKitchen"
        >
          <polygon
            points="50,0 100,0 100,100 50,100"
          />
        </g>
      </g>
    </svg>
  </div>
</div>
`;
